src/seismometer/data/loader/pipeline.py

```diff
@@ -0,0 +1,158 @@
+            and returning a dataframe.
+            Used to merge events onto predictions based on configuration.
+        """
+        self.config = config
+
+        self.prediction_fn = prediction_fn
+        self.post_predict_fn = post_predict_fn or _passthru_framehook
+
+        self.event_fn = event_fn
+        self.post_event_fn = post_event_fn or _passthru_framehook
+        self.merge_fn = merge_fn or _passthru_mergehook
+
+        self.prediction_from_memory: ConfigFrameHook = _passthru_framehook
+        self.event_from_memory: ConfigFrameHook = _passthru_framehook
+
+    def load_data(self, prediction_obj: pd.DataFrame = None, event_obj: pd.DataFrame = None) -> pd.DataFrame:
+        """
+        Entry point for loading data for a Seismogram session.
+
+        Can optionally accept a prediction and event frame that are prioritized over loading from configuration.
+
+        Parameters
+        ----------
+        prediction_obj : pd.DataFrame, optional
+            an optional pre-loaded predictions dataframe, if None then will load from configuration, by default None.
+        event_obj : pd.DataFrame, optional
+            an optional pre-loaded events dataframe, if None then will load from configuration, by default None.
+
+        Returns
+        -------
+        pd.DataFrame
+            the loaded and merged dataframe for a Seismogram session.
+        """
+
+        logger.info(f"Importing files from {self.config.config_dir}")
+
+        dataframe = self._load_predictions(prediction_obj)
+        dataframe = self.post_predict_fn(self.config, dataframe)
+        dataframe = self._add_events(dataframe, event_obj)
+
+        return dataframe
+
+    def _load_predictions(self, prediction_obj: pd.DataFrame = None):
+        """Load predictions from config or memory."""
+        if (prediction_obj is None) and (self.prediction_fn is None):
+            return pd.DataFrame()
+        if prediction_obj is None:
+            return self.prediction_fn(self.config)
+        return self.prediction_from_memory(self.config, prediction_obj)
+
+    def _add_events(self, dataframe: pd.DataFrame, event_obj: pd.DataFrame = None):
+        """Load and merge events onto the predictions dataframe."""
+        event_frame = self._load_events(event_obj)
+        if event_frame.empty:  # No events to add
+            return dataframe
+
+        event_frame = self.post_event_fn(self.config, event_frame)
+        return self.merge_fn(self.config, event_frame, dataframe)
+
+    def _load_events(self, event_obj: pd.DataFrame = None) -> pd.DataFrame:
+        """Load events from config or memory."""
+        if (event_obj is None) and (self.event_fn is None):
+            return pd.DataFrame()
+        if event_obj is None:
+            return self.event_fn(self.config)
+        return self.event_from_memory(self.config, event_obj)
+
+
+__all__ = ["SeismogramLoader", "ConfigOnlyHook", "ConfigFrameHook", "MergeHook"]
```

Comment:
Should this use the @export decorator?

Reply:
Three of the four exposed objects are typing generics and not functions.  

The decorator would need fairly significant reworking to handle these, _or_ we could override both __name\_\_ and __module\_\_ similar to the docstring.  
That felt much worse than just doing the manual declaration.